Add an Ambient mode documentation (#579)
istio-ecosystem/sail-operator: docs/common/istio-ambient-mode.md

```
    
Wait for the `bookinfo-gateway` pod running and then get the `productpage` service URL. The wait time depends on your cluster cloud provider. It takes me about one minute from an AWS ELB to be able to access it.

```bash
$ export INGRESS_HOST=$(kubectl get -n bookinfo gtw bookinfo-gateway -o jsonpath='{.status.addresses[0].value}')
$ export INGRESS_PORT=$(kubectl get -n bookinfo gtw bookinfo-gateway -o jsonpath='{.spec.listeners[?(@.name=="http")].port}')
$ export GATEWAY_URL=$INGRESS_HOST:$INGRESS_PORT
$ echo "http://${GATEWAY_URL}/productpage"
```

5. Access the application.

Open your browser and navigate to `http://${GATEWAY_URL}/productpage` to view the Bookinfo application.
If you refresh the page, you should see the display of the book ratings changing as the requests are distributed across the different versions of the reviews service.

6. Add Bookinfo to the Ambient mesh.

```bash
$ kubectl label namespace bookinfo istio.io/dataplane-mode=ambient
```

> **Note**:
> You don't need to restart or redeploy any of the application pods.

> > Unlike the sidecar mode, each pod's container count will remain one in the ambient mode.

Reviewer: Some pods might have more than one container in their typical deployments, so lets modify the sentence as....

```suggestion
> Unlike the sidecar mode, each pod's container count will remain the same even after adding them to the ambient mesh.
```

## Istio Ambient mode

> This documentation is specific to the Istio Ambient mode using Sail Opeartor installation. These features provide early access to upcoming product features, enabling users to test functionality and provide feedback during the development process.

Reviewer: I think we can remove this whole section as its NOT the product documentation.

## Introduction to Istio Ambient mode

Ambient mesh and its reference implementation with Istio's ambient mode was announced in September 2022. It reaches a stable status in Istio 1.24 release. The core innovation behind ambient mesh is that it slices Layer 4(L4) and Layer 7(L7) processing into two distinct layers. Istio's ambient mode is powered by lightweight, shared L4 node proxies and optional L7 proxies, removing the needfor traditional sidecar proxies from the data plane.

> The lightweight shared L4 node proxy is called the `ztunnel`(zero-trust tunnel). ztunnel drastically reduces the overhead of running a mesh by removing the need to potentially over-provision memory and CPU within a cluster to handle expected loads, while still providing security using mutual TLS with cryptographic identity, simple L4 authorization policies, and telemetry.

Reviewer: ```suggestion
The lightweight, shared L4 node proxy is called ztunnel (zero-trust tunnel). It significantly reduces the overhead of running a mesh by eliminating the need to over-provision memory and CPU while still ensuring security through mutual TLS, simple L4 authorization policies, and telemetry.
```

> The L7 proxies are called `waypoints`. Waypoints process L7 functions such as traffic routing, authorization policy enforcement, and resilience. Waypoints run outside of your application deployments and can scale independently based on your needs.

Reviewer: ```suggestion
The L7 proxies, known as waypoints, handle functions like traffic routing, authorization policy enforcement, and resilience. They run separately from application deployments and can scale independently based on demand.
```

In contrast to the earlier injection of sidecars. Users can start with the secure L4 overlay, which offers features such as mTLS, authorization policy, and telemetry. Complex L7 handling such as retries, traffic splitting, load balancing, and observability collection can then be enabled on a case-by-case basis.

## Component version

> For running Istio ambient mode on OpenShift, we recommend using Sail Operator and installing `Istio`, `IstioCNI` and `ZTunnel` resources version v1.24.0 or newer.

Reviewer: As this documentation is for Sail Operator, we can be generic. What's your though about the following sentence.

```suggestion
To install Istio Ambient mode using Sail Operator, use version `v1.24.0` or later and deploy the `Istio`, `IstioCNI`, and `ZTunnel` resources.
```

Author: I think the version `v1.24.0` or later is only required when installing it on OpenShift. I will rephrase it as :

```
To install Istio Ambient mode using Sail Operator on OpenShift, use version `v1.24.0` or later and deploy the `Istio`, `IstioCNI`, and `ZTunnel` resources.
```

## Concepts

### ZTunnel resource

> The `ZTunnel` resource is used to manage your L4 node proxy. It is a cluster-wide resource as it will install a `DaemonSet` that will be operating on all nodes of your cluster. You can select a version by setting the `spec.version` field, as you can see in the sample below. Just like the `Istio` resource, it also has a `values` field that exposes all of the options provided in the `ztunnel` chart:

Reviewer: ```suggestion
The `ZTunnel` resource manages the L4 node proxy and is a cluster-wide resource. It deploys a DaemonSet that runs on all nodes in the cluster. You can specify the version using the `spec.version` field, as shown in the example below. Similar to the `Istio` resource, it also includes a `values` field that allows you to configure options available in the ztunnel helm chart.
```

```yaml
apiVersion: sailoperator.io/v1alpha1
kind: ZTunnel
metadata:
  name: default
spec:
  profile: ambient
  version: v1.24.0
  namespace: ztunnel
  values:
    ztunnel:

>       image: docker.io/istio/ztunnel:1.24.0

Reviewer: Instead of using `image` in the example (which is matching with the version), shall we use logging?
I'm fine, if you prefer to retain the image example if its a more widely used use-case.

```
values:
  logging:
    level: "all:debug"
```

Author: ok, I will keep the image example there because we may switch to other image definition in the future.

```

### API Reference documentation

The ZTunnel resource API reference documentation can be found [here](https://github.com/istio-ecosystem/sail-operator/blob/main/docs/api-reference/sailoperator.io.md#ztunnel).

## Core features

> - Secure application access using mTLS encryption.
> - Secure application access using Layer 4 authorization policies.
> - Enfore Layer 7 authorization policies using a waypoint proxy.
> - Split traffic between services using a waypoint proxy.
> - Gathering TCP telemetry and visualize all traffic between the pods.

Reviewer: ```suggestion
- Secure application access using mTLS encryption.
- Enforce Layer 4 authorization policies for enhanced security.
- Apply Layer 7 authorization policies using a waypoint proxy.
- Route and split traffic between services with a waypoint proxy.
- Collect TCP telemetry and visualize traffic flow between pods using Kiali.
```

## Getting Started

### Installation on OpenShift

*Prerequisites*

* You have access to the cluster as a user with the `cluster-admin` cluster role.

*Steps*

> 1. Install a Sail Operator using the CLI or through the web console. The steps can be found [here](https://github.com/istio-ecosystem/sail-operator/blob/main/docs/README.md#installation-on-openshift).

Reviewer: ```suggestion
1. Install the Sail Operator using the CLI or through the web console. The steps can be found [here](https://github.com/istio-ecosystem/sail-operator/blob/main/docs/README.md#installation-on-openshift).
```

    Note: `IN USE` field shows as 0, as `Istio` has just been installed and there are no workloads using it.

5. Create the `istio-cni` namespace.

```bash
$ kubectl create namespace istio-cni
```

6. Create the `IstioCNI` resource.

```bash
$ cat <<EOF | kubectl apply -f-
apiVersion: sailoperator.io/v1alpha1
kind: IstioCNI
metadata:
  name: default
spec:
  profile: ambient
  version: v1.24.0
  namespace: istio-cni
EOF
```

7. Create the `ztunnel` namespace and add a label `istio-discovery=enabled`.
> **NOTE**: 
> We need to label both the `Istio` resource's namespace e.g. `istio-system` and the `ZTunnel` resource's namespace when using a `discoverySelectors` mesh config. Those two labels should be added before installing a `ZTunnel` instance. This approach is used to avoid a [TLS signing error](https://github.com/istio/istio/issues/52057).

```bash
$ kubectl create namespace ztunnel
$ kubectl label namespace ztunnel istio-discovery=enabled
```

8. Create the `ZTunnel` resource.

```bash
$ cat <<EOF | kubectl apply -f-
apiVersion: sailoperator.io/v1alpha1
kind: ZTunnel
metadata:
  name: default
spec:
  profile: ambient
  version: v1.24.0
  namespace: ztunnel
  values:
    ztunnel:

>       image: docker.io/istio/ztunnel:1.24.0

Reviewer: Probably we can remove this section as the default image pointing to the version will be used anyways.

EOF
```

9. Confirm the installation and version of the `ztunnel`.

```console
$ kubectl get ztunnel -n istio-system
    NAME      READY   STATUS    VERSION   AGE
    default   True    Healthy   v1.24.0   16s
```

### Deploy a sample application

> To explore Istio's ambient mode, you will install the sample `Booinfo application`.

Reviewer: ```suggestion
To explore Istio's ambient mode, lets install the sample `Bookinfo application`.
```

$ kubectl apply -n bookinfo -f https://raw.githubusercontent.com/istio/istio/release-1.24/samples/bookinfo/gateway-api/bookinfo-gateway.yaml
```
    
Wait for the `bookinfo-gateway` pod running and then get the `productpage` service URL. The wait time depends on your cluster cloud provider. It takes me about one minute from an AWS ELB to be able to access it.

```bash
$ export INGRESS_HOST=$(kubectl get -n bookinfo gtw bookinfo-gateway -o jsonpath='{.status.addresses[0].value}')
$ export INGRESS_PORT=$(kubectl get -n bookinfo gtw bookinfo-gateway -o jsonpath='{.spec.listeners[?(@.name=="http")].port}')
$ export GATEWAY_URL=$INGRESS_HOST:$INGRESS_PORT
$ echo "http://${GATEWAY_URL}/productpage"
```

5. Access the application.

Open your browser and navigate to `http://${GATEWAY_URL}/productpage` to view the Bookinfo application.
If you refresh the page, you should see the display of the book ratings changing as the requests are distributed across the different versions of the reviews service.

6. Add Bookinfo to the Ambient mesh.

```bash
$ kubectl label namespace bookinfo istio.io/dataplane-mode=ambient
```

> If you refresh the previous browser page, you should see the same display.

Reviewer: Shall we add a note here saying that there is no need to restart any of the bookinfo pods and also the containers count will continue remain one (unlike the sideCar mode where we would see two containers).

## Visualize the application using Kiali dashboard

Using Kiali dashboard and Prometheus metrics engine, you can visualize the Bookinfo application traffic and mTLS encryption.

> Deploy a Prometheus in `istio-system` namespace.

Reviewer: ```suggestion
Deploy Prometheus in `istio-system` namespace.
```

```bash
$ kubectl apply -n istio-system -f https://raw.githubusercontent.com/istio/istio/master/samples/addons/prometheus.yaml
```

Deploy a Kiali dashboard using a community Kiali operator on OpenShift.

```bash
$ cat <<EOF | kubectl apply -f -
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: kiali
  namespace: openshift-operators
spec:
  channel: stable
  installPlanApproval: Automatic
  name: kiali
  source: community-operators
  sourceNamespace: openshift-marketplace
EOF
$ kubectl wait --for condition=established --timeout=60s crd "kialis.kiali.io"

customresourcedefinition.apiextensions.k8s.io/kialis.kiali.io condition met

$ cat <<EOF | kubectl apply -f -
apiVersion: kiali.io/v1alpha1
kind: Kiali
metadata:
  name: kiali
  namespace: istio-system
EOF
```

> To access the Kiali dashboard, you will get the URL below.

Reviewer: ```suggestion
To access the Kiali dashboard, lets get the URL.
```

- [Troubleshoot connectivity issues with ztunnel](https://istio.io/latest/docs/ambient/usage/troubleshoot-ztunnel/)

Before adding Bookinfo to the Ambient mesh, you should get all workloads using TCP protocol.

```bash
$ istioctl -n ztunnel ztunnel-config workloads
    NAMESPACE    POD NAME                        ADDRESS      NODE                       WAYPOINT PROTOCOL
    bookinfo     details-v1-6cd6d9df6b-mddv2     10.129.0.43  ip-10-0-0-241.ec2.internal None     TCP
    bookinfo     productpage-v1-57ffb6658c-vmb6z 10.129.0.48  ip-10-0-0-241.ec2.internal None     TCP
    bookinfo     ratings-v1-794744f5fd-wm798     10.129.0.44  ip-10-0-0-241.ec2.internal None     TCP
    bookinfo     reviews-v1-67896867f4-4h9j6     10.129.0.45  ip-10-0-0-241.ec2.internal None     TCP
    bookinfo     reviews-v2-86d5db4bd6-dbw4f     10.129.0.46  ip-10-0-0-241.ec2.internal None     TCP
    bookinfo     reviews-v3-77947c4c78-r54c9     10.129.0.47  ip-10-0-0-241.ec2.internal None     TCP
...
```

> After adding it to the Ambient mesh, you should see a different result using HBONE protocol.

Reviewer: ```suggestion
After adding `bookinfo` namespace to the Ambient mesh, you would see HBONE protocol..
```

...
```

- [Verify mutual TLS is enabled](https://istio.io/latest/docs/ambient/usage/verify-mtls-enabled/)

You can also validate mTLS from ztunnel logs to confirm mTLS is enabled.

```bash
$ kubectl -n ztunnel logs -l app=ztunnel | grep -E "inbound|outbound"

2025-01-23T05:07:25.806642Z	info	access	connection complete	src.addr=10.129.0.48:40978 src.workload="productpage-v1-57ffb6658c-vmb6z" src.namespace="bookinfo" src.identity="spiffe://cluster.local/ns/bookinfo/sa/bookinfo-productpage" dst.addr=10.129.0.43:15008 dst.hbone_addr=10.129.0.43:9080 dst.service="details.bookinfo.svc.cluster.local" dst.workload="details-v1-6cd6d9df6b-mddv2" dst.namespace="bookinfo" dst.identity="spiffe://cluster.local/ns/bookinfo/sa/bookinfo-details" direction="outbound" bytes_sent=283 bytes_recv=358 duration="2ms"
```

Validate the `src.identity` and `dst.identity` values are correct. They are the identities used for the mTLS communication among the source and destination workloads.

### Cleanup

If you no longer need associated resources, you can delete them by following the steps below.

#### Remove waypoint proxies

```bash
kubectl label namespace bookinfo istio.io/use-waypoint-

> kubectl delete -n bookinfo Gateway waypoint

Reviewer: Is this required as we are deleting the bookinfo-gateway in L347?

Author: no, I will move this part in my waypoint doc file later.

```

#### Remove the namespace from the ambient data plane

```bash
kubectl label namespace bookinfo istio.io/dataplane-mode-
```

> **NOTE**:

> > You must remove workloads from the ambient data plane before uninstalling Istio.

Reviewer: Did you notice any issues when Istio is uninstalled without removing the workloads from ambient?

Author: I haven't tried that scenario. This note is mentioned from upstream doc so I copied it from upstream doc:

https://istio.io/latest/docs/ambient/getting-started/cleanup/#remove-the-namespace-from-the-ambient-data-plane

## Introduction to Istio Ambient mode

Ambient mesh was announced in September 2022 and reached a stable state in the Istio 1.24 release. It's core innovation is the separation of Layer 4 (L4) and Layer 7 (L7) processing into two distinct layers. It uses lightweight, shared L4 node proxies and optional L7 proxies, eliminating the need for traditional sidecar proxies in the data plane.

The lightweight, shared L4 node proxy is called ztunnel (zero-trust tunnel). It significantly reduces the overhead of running a mesh by eliminating the need to over-provision memory and CPU while still ensuring security through mutual TLS, simple L4 authorization policies, and telemetry.

The L7 proxies, known as waypoints, handle functions like traffic routing, authorization policy enforcement, and resilience. They run separately from application deployments and can scale independently based on demand.

> In contrast to the Istio Sidecar mode. Users can start with the secure L4 overlay, which offers features such as mTLS, authorization policy, and telemetry. Complex L7 handling such as retries, traffic splitting, load balancing, and observability collection can then be enabled on a case-by-case basis.

Reviewer 2: Hi @yxun I still don't understand this sentence. "In contrast to the Istio Sidecar mode. Users can start with the secure L4 overlay..."

Reviewer: Hello @nyechiel, @yxun is on PTO for the rest of the week, so let me try if I can clarify it. 

The main difference from Istio’s Sidecar mode is that with Ambient mode, we don’t have to enable full Layer 7 features right away. We can start with a lightweight, secure Layer 4 overlay, which gives us features like mTLS, authorization policies, and telemetry. If we need more advanced L7 features, like retries, traffic splitting etc, we can create waypoints as needed. In Sidecar mode, L7 processing is always included by default, so we don’t have this flexibility.

Please feel free to suggest/comment on any improvements. Thank you.

Reviewer: Oh yeah, good catch! There should be a "," instead of "."

Author: I will include this change in another PR when I finish a waypoint user guide draft early this week.

Author: this issue is addressed in PR https://github.com/istio-ecosystem/sail-operator/pull/694

## Introduction to Istio Ambient mode

Ambient mesh was announced in September 2022 and reached a stable state in the Istio 1.24 release. It's core innovation is the separation of Layer 4 (L4) and Layer 7 (L7) processing into two distinct layers. It uses lightweight, shared L4 node proxies and optional L7 proxies, eliminating the need for traditional sidecar proxies in the data plane.

The lightweight, shared L4 node proxy is called ztunnel (zero-trust tunnel). It significantly reduces the overhead of running a mesh by eliminating the need to over-provision memory and CPU while still ensuring security through mutual TLS, simple L4 authorization policies, and telemetry.

The L7 proxies, known as waypoints, handle functions like traffic routing, authorization policy enforcement, and resilience. They run separately from application deployments and can scale independently based on demand.

> In contrast to the earlier injection of sidecars. Users can start with the secure L4 overlay, which offers features such as mTLS, authorization policy, and telemetry. Complex L7 handling such as retries, traffic splitting, load balancing, and observability collection can then be enabled on a case-by-case basis.

Reviewer 2: Something seems to be wrong with this line ("In contrast to the earlier injection of sidecars."). Perhaps it should be part of the previous sentence?

Author: Hello Nir, The word `earlier` is confusing here. This sentence is not related to the previous L7 waypoint features. It compares the Ambient mode with the Istio Sidecar mode. I will rephrase it.

customresourcedefinition.apiextensions.k8s.io/kialis.kiali.io condition met

$ cat <<EOF | kubectl apply -f -
apiVersion: kiali.io/v1alpha1
kind: Kiali
metadata:
  name: kiali
  namespace: istio-system
EOF
```

To access the Kiali dashboard, let's get the URL.

```bash
$ kubectl get route -n istio-system -l app.kubernetes.io/name=kiali -o jsonpath='https://{..spec.host}/'
```

![kiali-dashboard](images/kiali-dashboard.png "Kiali Dashboard")

Send some traffic to the Bookinfo application and open the Kiali dashboard page. Click on the Traffic Graph and select `bookinfo` from the `Select Namespaces` drop-down. You should see the Bookinfo application traffic flow in the graph.

![kiali-traffic-map](images/kiali-traffic-map.png "Kiali Traffic Map")

Next, click and select `Show Badges`, `Security` from the `Display` drop-down. You should see each Bookinfo application traffic edge with a lock icon. By default, the traffic between services is mTLS encrypted in Istio ambient mode.

![kiali-security](images/kiali-security.png "Kiali Security")

In the [Kiali documentation](https://kiali.io/docs/features/ambient/) there is a list of all the Ambient features. 

### Troubleshoot issues

> A brief and helpful troubleshooting guide can be reviewed from the upstream Istio documentation.

Reviewer 2: Any reason not to link the URL directly here?
